olp-cpp-sdk-dataservice-read/src/repositories/QuadTreeIndex.cpp

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright (C) 2020 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+#include "QuadTreeIndex.h"
+
+#include <rapidjson/document.h>
+#include <algorithm>
+#include <bitset>
+#include <iostream>
+#include "generated/model/Index.h"
+// clang-format off
+#include "generated/parser/IndexParser.h"
+#include "generated/parser/PartitionsParser.h"
+#include <olp/core/generated/parser/JsonParser.h>
+// clang-format on
+
+namespace olp {
+namespace dataservice {
+namespace read {
+
+QuadTreeIndex::LayerIndex QuadTreeIndex::FromJson(const olp::geo::TileKey& root,
+                                                  int depth,
+                                                  const std::string& json) {
+  std::vector<LayerIndex::IndexData> subs;
+  std::vector<LayerIndex::IndexData> parents;
+
+  model::Index idx = olp::parser::parse<model::Index>(json);
+  // TODO reduce copying from index
+  for (const auto& subQuad : idx.GetSubQuads()) {
+    LayerIndex::IndexData record;
+    record.tileKey_ = root.AddedSubHereTile(subQuad->GetSubQuadKey());
+    record.data_handle_ = subQuad->GetDataHandle();
+    record.version_ = subQuad->GetVersion();
+    subs.push_back(std::move(record));
+  }
+
+  std::sort(subs.begin(), subs.end());
+
+  for (auto& parentQuad : idx.GetParentQuads()) {
+    LayerIndex::IndexData record;
+    record.tileKey_ = root.FromQuadKey(parentQuad->GetPartition());
```

Comment:
Is this correct? Should be

```suggestion
    record.tileKey_ = root.FromHereTile(parentQuad->GetPartition());
```